test.js

```diff
@@ -0,0 +1,191 @@
+    });
+
+    it('should return false with zero-length Buffer values', () => {
+      assume(subject(Buffer.of(0), Buffer.of(0), Buffer.of(0))).is.not.ok();
+    });
+
+    it('should return false with all new-line string values', () => {
+      assume(subject('\n\n\n', '\n\n\n', '\n\n\n')).is.not.ok();
+    });
+
+    it('should return false with all new-line buffer values', () => {
+      assume(subject(Buffer.from('\n\n\n'), Buffer.from('\n\n\n'), Buffer.from('\n\n\n'))).is.not.ok();
+    });
+
+    it('should throw error with invalid cert structured data', () => {
+      assume(() => {
+        subject(pubkey, '', 'kaldsjflasjflsdf');
+      }).throws(/PEM_ASN1_read_bio/);
+    });
+
+    it('should throw error with invalid structured data', () => {
+      assume(() => {
+        subject('kadjflakdjfa', '', pkcs7);
+      }).throws(/PEM_read_bio/);
+    });
+  });
+
+  describe('with valid files', () => {
+    it('should validate valid credentials with header in buffer', () => {
+      assume(subject(pubkey, document, pkcs7)).is.ok();
+    });
+
+    it('should validate valid credentials without header in buffer', () => {
+      assume(subject(pubkey, document, pkcs7_with_header)).is.ok();
+    });
+
+    it('should validate valid credentials with header in string', () => {
+      assume(subject(pubkey.toString(), document.toString(), pkcs7.toString())).is.ok();
+    });
+
+    it('should validate valid credentials without header in string', () => {
+      assume(subject(pubkey.toString(), document.toString(), pkcs7_with_header.toString())).is.ok();
+    });
+  });
+
+  // Slightly, as in off by a very small amount
+  describe('with slightly invalid files', () => {
+    it('should fail to validate with an extra character on document', () => {
+      for (let i = 0; i <= 255; i++) {
+        let badDoc = Buffer.concat([
+          document,
+          Buffer.from([i]),
+        ]);
+        assume(subject(pubkey, badDoc, pkcs7_with_header)).is.not.ok();
+      }
+    });
+
+    // A bit flip on any bit of this file should generate an error.  We know this
+    it('should fail for bitflips on document', () => {
+      for (let i = 0; i < document.length; i++) {
+        for (let j = 0; j < 8; j++) {
+          let badDoc = Buffer.from(document);
+          /*console.log([
+            'Byte ', i,
+            'Bit ', j,
+            parseInt(badDoc[i], 10).toString(2),
+            ' -> ',
+            parseInt(badDoc[i] ^ (1 << j), 10).toString(2),
+          ].join(' '))*/
+          badDoc[i] ^= 1 << j;
+          assume(subject(pubkey, badDoc, pkcs7_with_header)).is.not.ok();
+        }
+      }
+    });
+
+    // We're going to flip some bits here.  We're only going to try to flip the
+    // bits that relate to the cryptographic calculations.  We don't flip all
+    // bits for this document since we're not doing signature file validation
+    // because of a self-signed certificate being used (i.g. -noverify)
```

Comment:
this is really cool!

Reply:
thanks :)  It was a fun set of tests to write